cabal-install/src/Distribution/Client/CmdRepl.hs

```diff
@@ -285,15 +287,56 @@ multiReplDecision ctx compiler flags =
 -- For more details on how this works, see the module
 -- "Distribution.Client.ProjectOrchestration"
 replAction :: NixStyleFlags ReplFlags -> [String] -> GlobalFlags -> IO ()
-replAction flags@NixStyleFlags{extraFlags = r@ReplFlags{..}, ..} targetStrings globalFlags =
-  withContextAndSelectors verbosity AcceptNoTargets (Just LibKind) flags targetStrings globalFlags ReplCommand $ \targetCtx ctx targetSelectors -> do
+replAction flags@NixStyleFlags{extraFlags = r@ReplFlags{..}, ..} targetStrings' globalFlags = do
+  -- NOTE: The REPL will work with no targets in the context of a project if a
+  -- single package is in the same directory as the project file. To have the
+  -- same behaviour when the package is somewhere else we adjust the targets.
+  targetStrings <-
+    if null targetStrings'
+      then withCtx silent targetStrings' $ \targetCtx ctx _ ->
+        return . fromMaybe [] $ case targetCtx of
+          ProjectContext ->
+            let pkgs = projectPackages $ projectConfig ctx
+             in if length pkgs == 1
+                  then pure <$> listToMaybe pkgs
+                  else Nothing
+          _ -> Nothing
+      else return targetStrings'
+
+  withCtx verbosity targetStrings $ \targetCtx ctx targetSelectors -> do
     when (buildSettingOnlyDeps (buildSettings ctx)) $
       dieWithException verbosity ReplCommandDoesn'tSupport
     let projectRoot = distProjectRootDirectory $ distDirLayout ctx
         distDir = distDirectory $ distDirLayout ctx
 
     baseCtx <- case targetCtx of
-      ProjectContext -> return ctx
+      ProjectContext -> do
+        let pkgs = projectPackages $ projectConfig ctx
+        when (null targetStrings && length pkgs /= 1) $
+          let projectName = case projectConfigProjectFile . projectConfigShared $ projectConfig ctx of
+                Flag "" -> Nothing
+                Flag n -> Just $ quotes (text n)
+                _ -> Nothing
+              msg =
+                case (null pkgs, projectName) of
+                  (True, Just project) ->
+                    text "There are no packages in"
+                      <+> (project <> char '.')
+                      <+> text "Please add a package to the project and pick a component to use as the target of the REPL command."
```

Comment:
the no-packages case is certainly already handled in other places today, right? (people keep forgetting `packages:...` and get an error message.) Is it possible to not duplicate this logic?

Reply:
> Warning: There are no packages or optional-packages in the project

is a different issue, I think, (https://github.com/haskell/cabal/issues/8679 probably) so it may be wise to try to avoid solving it here.

Reply:
The assertion bothers me. I could cut back on the scope of this pull request or dig in more about the assertion.

Reply:
Given the warning (unless there's already some way to suppress it, shouldn't the error message change to be a follow-on to it?